src/storage/src/source/mysql/replication.rs

```diff
@@ -0,0 +1,405 @@
+                return Ok(());
+            }
+
+            let connection_config = connection
+                .connection
+                .config(&*config.config.connection_context.secrets_reader, &config.config)
+                .await?;
+
+            let mut conn = connection_config.connect(
+                &format!("timely-{worker_id} MySQL replication reader"),
+                &config.config.connection_context.ssh_tunnel_manager,
+            ).await?;
+
+            // Get the set of GTIDs currently available in the binlogs (GTIDs that we can safely start replication from)
+            let binlog_gtid_set: String = conn.query_first("SELECT GTID_SUBTRACT(@@global.gtid_executed, @@global.gtid_purged)").await?.unwrap();
+            // NOTE: The assumption here for the MVP is that there is only one source-id in the GTID set, so we can just
+            // take the transaction_id from the first one. Fix this once we move to a partitioned timestamp.
+            let binlog_start_gtid = GtidSet::from_str(&binlog_gtid_set)?.first().expect("At least one gtid").to_owned();
+            let binlog_start_transaction_id = TransactionId::new(binlog_start_gtid.earliest_transaction_id());
+
+            let resume_upper = Antichain::from_iter(
+                subsource_resume_uppers
+                    .values()
+                    .flat_map(|f| f.elements())
+                    // Advance any upper as far as the start of the binlog.
+                    .map(|t| std::cmp::max(*t, binlog_start_transaction_id))
+            );
+
+            let Some(resume_transaction_id) = resume_upper.into_option() else {
+                return Ok(());
+            };
+            data_cap_set.downgrade([&resume_transaction_id]);
+            upper_cap_set.downgrade([&resume_transaction_id]);
+            trace!(%id, "timely-{worker_id} replication reader started transaction_id={}", resume_transaction_id);
+
+            let mut rewinds = BTreeMap::new();
+            while let Some(event) = rewind_input.next().await {
+                if let AsyncEvent::Data(caps, data) = event {
+                    for req in data {
+                        let snapshot_transaction_id = TransactionId::new(req.snapshot_gtid_set.first().expect("at least one gtid").latest_transaction_id());
+                        if resume_transaction_id > snapshot_transaction_id + TransactionId::new(1) {
+                            let err = DefiniteError::BinlogCompactedPastResumePoint(resume_transaction_id.to_string(), (snapshot_transaction_id + TransactionId::new(1)).to_string());
+                            // If the replication stream cannot be obtained from the resume point there is
+                            // nothing else to do. These errors are not retractable.
+                            for (output_index, _) in table_info.values() {
+                                // We pick `u64::MAX` as the TransactionId which will (in practice) never conflict
```

Comment:
I think we should add an associated constant to `TransactionId` itself so that we can do `TransactionId::MAX` in all places where we have this logic here. If we end up changing the representation in `encode_row` to `Int64` then the `MAX` constant should change to `i64::MAX` and it sounds better to update this in a single place instead of hunting down all the places that attempt a `u64::MAX` and fixing them